src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -1703,7 +1707,7 @@ abstract class RefChecks extends Transform {
               assert(sym != NoSymbol, "transformCaseApply: name = " + name.debugString + " tree = " + tree + " / " + tree.getClass) //debug
               enterReference(tree.pos, sym)
             }
-            tree
+            eliminatedRightAssocValDefs.getOrElse(tree.symbol, tree)
```

Comment:
In general we cannot just move trees around like this, because the owner chain of symbols defined in the tree might mess up. This is maybe what's causing the following crash:
```
scala> class C { def f_:(x: => Int) = 0 }
defined class C

scala> val c = new C
c: C = C@4dd94931

scala> { val x = 1; x } f_: c
java.util.NoSuchElementException: key not found: value x
	at scala.collection.MapLike.default(MapLike.scala:230)
	at scala.collection.MapLike.default$(MapLike.scala:229)
	at scala.collection.AbstractMap.default(Map.scala:59)
	at scala.collection.mutable.HashMap.apply(HashMap.scala:61)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder$locals$.load(BCodeSkelBuilder.scala:391)
```